jcl/src/openj9.dtfj/share/classes/com/ibm/dtfj/javacore/parser/j9/section/threadhistory/ThreadHistorySectionParser.java

```diff
@@ -0,0 +1,157 @@
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] https://openjdk.org/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+package com.ibm.dtfj.javacore.parser.j9.section.threadhistory;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import com.ibm.dtfj.image.CorruptDataException;
+import com.ibm.dtfj.image.DataUnavailable;
+import com.ibm.dtfj.javacore.builder.IImageAddressSpaceBuilder;
+import com.ibm.dtfj.javacore.builder.IImageProcessBuilder;
+import com.ibm.dtfj.javacore.builder.IJavaRuntimeBuilder;
+import com.ibm.dtfj.javacore.builder.javacore.ImageBuilder;
+import com.ibm.dtfj.javacore.builder.javacore.JavaRuntimeBuilder;
+import com.ibm.dtfj.javacore.parser.framework.parser.ParserException;
+import com.ibm.dtfj.javacore.parser.j9.IAttributeValueMap;
+import com.ibm.dtfj.javacore.parser.j9.SectionParser;
+import com.ibm.dtfj.javacore.parser.j9.section.common.TraceTimeParser;
+
+/**
+ * Provides parser for Thread history section in the javacore.
+ */
+public class ThreadHistorySectionParser extends SectionParser implements IThreadHistoryTypes {
+
+	private IImageAddressSpaceBuilder fImageAddressSpaceBuilder;
+	private IImageProcessBuilder fImageProcessBuilder;
+	private IJavaRuntimeBuilder fJavaRuntimeBuilder;
+
+	/**
+	 * Construct a javacore parser for the trace
+	 * of the current thread.
+	 */
+	public ThreadHistorySectionParser() {
+		super(THREADHISTORY_SECTION);
+	}
+
+	/**
+	 * Overall controls of parsing for the thread history section.
+	 * @throws ParserException on failure to parse the history due to IOException or excessive line length
+	 */
+	@Override
+	protected void topLevelRule() throws ParserException {
+		fImageAddressSpaceBuilder = fImageBuilder.getCurrentAddressSpaceBuilder();
+		fImageProcessBuilder = fImageAddressSpaceBuilder.getCurrentImageProcessBuilder();
+		fJavaRuntimeBuilder = fImageProcessBuilder.getCurrentJavaRuntimeBuilder();
+
+		parseThreadHistoryLine();
+	}
+
+	/**
+	 * Parse the thread history information.
+	 * Also uses the times in the thread history to fix the JVM start time
+	 * and the Javacore creation time which have been parsed elsewhere
+	 * with an ambiguous time zone.
+	 * The start time and creation times do not have a time zone, so have been parsed according
+	 * to the parser's time zone, which might differ from the time zone when creating the Javacore.
```

Comment:
Perhaps a better place to do this would be in `TitleSectionParser`; this section need not exist.

Reply:
We can't do this in TitleSectionParser. The title section has the creation time of the Javacore, but pre issue #12397 it has an unknown timezone, so is parsed in the timezone of the program running DTFJ, which might be different from the timezone of the target process. The title section parser cannot see later sections.

Later in the Javacore we can see some times (GC & thread trace) which are in UTC and presume these are close to, but before the writing of the Javacore. We then use those times to adjust the hours (and minutes in 15 minute steps) of the Javacore time. If the GC & thread trace sections are missing, no correction of the Javacore time can be made.